control task group concurrency: use allow list for mapping index

mis98zb opened the pull request:
To control the concurrency of task group, the option max_active_tis_per_dag and max_active_tis_per_dagrun on task_instance do not work well.

Below is my simplified case:

```python
import random
from datetime import timedelta
import logging

import pendulum
from airflow import DAG
from airflow.decorators import task, task_group
from airflow.sensors.base import PokeReturnValue

log = logging.getLogger(__name__)


@task(task_id='prepareJobParameter')
def prepare_job_parameter(ti=None) -> []:
    log.info('prepare script~~~~~~~~~~~~~~~~')
    return [
        {'parameter': 'aaaaaaaaaaaaaaa'},
        {'parameter': 'bbbbbbbbbbbbbbb'},
        {'parameter': 'ccccccccccccccc'}
    ]


@task_group(group_id=f'runScopeJob')
def run_job(task_name: str, submit_param: dict, check_interval: int, timeout: timedelta) -> None:
    @task(task_id=f'submitJob')
    def submit_job(job_param: dict, ti=None) -> str:
        log.info('submit job~~~~~~~~~~~~~~~~')
        return 'job url'

    @task.sensor(task_id=f'waitJob', poke_interval=check_interval, timeout=timeout.total_seconds(), mode="reschedule")
    def wait_job(job_url: str, ti=None) -> PokeReturnValue:
        log.info('check job status~~~~~~~~~~~~~~~~')
        status = random.randint(0, 3)
        if status > 0:
            return PokeReturnValue(is_done=False, xcom_value=status)
        else:
            return PokeReturnValue(is_done=True, xcom_value=status)

    wait_job(submit_job(submit_param))


@task(task_id="publishData")
def publish_data(ti=None) -> bool:
    log.info('publish data~~~~~~~~~~~~~~~~')
    return True


with DAG(
        dag_id='test_task_group_parallel',
        catchup=False,
        start_date=pendulum.datetime(2023, 4, 15, tz="UTC"),
        schedule=timedelta(hours=4),
        tags=['igs pipeline']
) as pipeline:
    task_pre_job = prepare_job_parameter()
    task_run_job = run_job.partial(task_name='Bingen', check_interval=180, timeout=timedelta(hours=24)).expand(
        submit_param=task_pre_job)
    task_run_job >> publish_data()


if __name__ == "__main__":
    pipeline.test()
```

Inside the mapped task group, there are two tasks: submit_job and wait_job.
The submit_job will finish very quickly.
The wait_job will be rescheduled till the job complete.
The max_active_tis_per_dagrun on submit_job cannot control the concurrency of the job, because it finished quickly.
All jobs will be submitted in a short time.

Commit:
control task group concurrency: use allow list for mapping index.

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

uranusjr commented on airflow/decorators/task_group.py:
```diff
@@ -124,14 +124,14 @@ def partial(self, **kwargs: Any) -> _TaskGroupFactory[FParams, FReturn]:
         return attr.evolve(self, partial_kwargs=kwargs)  # type: ignore[arg-type]
 
-    def expand(self, **kwargs: OperatorExpandArgument) -> DAGNode:
+    def expand(self, concurrency_limit: int = None, **kwargs: OperatorExpandArgument) -> DAGNode:
```
Should this be a `partial()` argument instead? It seems quite wrong to have an argument in `expand()` that does not actually expand. It would also be awkward to implement in `expand_kwargs()`.

mis98zb replied:
Got it, I'll move the parameter to partial().

I previously thought about this.
At that time, my understand is the partial() pass the common parameter to mapped operators, and expand() pass the parameter for expanding.
So I put the expanding concurrency limitation to expand()...

mis98zb replied:
@uranusjr I got a little confused when I move parameter to `partial()`.

Below is the partial() definition in decorators/task_group.py:
```python
    def partial(self, **kwargs: Any) -> _TaskGroupFactory[FParams, FReturn]:
        self._validate_arg_names("partial", kwargs)
        prevent_duplicates(self.partial_kwargs, kwargs, fail_reason="duplicate partial")
        kwargs.update(self.partial_kwargs)
        # TODO: fixme when mypy gets compatible with new attrs
        return attr.evolve(self, partial_kwargs=kwargs) 
```

While _validate_arg_names(), only the parameter for task group function is allow for kwargs.
For example, my test code is:
```python
@task_group
def tg(who: str):
    tgt1 = BashOperator(task_id="dummy1", bash_command="sleep 300")
    tgt2 = BashOperator(task_id="dummy2", bash_command="echo done")
    tgt1 >> tgt2
```
So only parameter 'who' is allowed for kwargs.
Seems the partial() is designed to pass parameters to task group function, not for expand execution... 

So I think expand() may be still the best place to add concurrency limit parameter.

How do you think about this?

mis98zb replied:
btw, airflow code is not compatible with pydantic 2.x, which blocked me a long time for debugging in local machine, T___T

uranusjr replied:
Thinking this again, do this in _neither_ `partial` and `expand` makes sense, since those target the parameters of the _decorated function_, not _task_ parameters. The best way to do it is likely in the `@task_group` decorator—yes this makes the argument appear when the task is not mapped, but in the case it can simply be a no-op, which should be fine as long as it is appropriately documented. This would also make it possible to use the argument in non-decorator version of the task group.

mis98zb replied:
Good idea!
This is consistent with the usage of max_active_tis_per_dagrun for task. 

So is it better to use the parameter name as `max_active_tgs_per_dagrun`?
tgs may be not so good, but I did not find any word better...

uranusjr replied:
Maybe `groups`? It’s pretty obviously task groups, after all.

mis98zb replied:
I have a question for decorated function.

I predefined some common flow as task group:
```python
@task_group(group_id=f'runJob')
def run_job(task_name: str, submit_param: dict, check_interval: int, timeout: timedelta) -> None:
    @task(task_id=f'submit{task_name}Job')
    def submit_scope_job(job_param: dict, ti=None) -> str:
        # submit job....
        return job_url

    @task.sensor(task_id=f'wait{task_name}Job', poke_interval=check_interval, timeout=timeout.total_seconds(), mode="reschedule")
    def wait_job(job_url: str, ti=None) -> PokeReturnValue:
        # query job status....
        if status == 'running':
            return PokeReturnValue(is_done=False, xcom_value=status)
        elif status == 'suceeded':
            return PokeReturnValue(is_done=True, xcom_value=status)
        elif status == 'failed':
            raise AirflowFailException(f'Job {job_url} failed with status {status}')

    wait_scope_job(submit_scope_job(submit_param))
```
And use it in may place as `run_job.expand(xxxx)`.

If I put `max_active_groups_per_dagrun` as a parameters of the decorated function, what's the graceful way to pass the value to the decorator?

uranusjr replied:
Either just put it in `@task_group(...)`, or use `.override()`

mis98zb replied:
got it, thank you!

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

Commit:
change to use max_active_groups_per_dagrun on task_group.

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

Commit:
remove duplicate field in MapppedTaskGroup.

Commit:
fix style for static checking.

Commit:
fix style for static checking.

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -277,9 +284,44 @@ def __get_concurrency_maps(self, states: Iterable[TaskInstanceState], session: S
+            )
+            .filter(TI.map_index >= 0)
+            .order_by(TI.dag_id, TI.run_id, TI.map_index)
         )
```
```suggestion
        tg_concurrency_query: Iterator[tuple[str, str, str, int]] = session.execute(
            select(TI.task_id, TI.run_id, TI.dag_id, TI.map_index)
            .where(
                TI.map_index >= 0,
                TI.state.in_(
                    (
                        TaskInstanceState.SCHEDULED,
                        TaskInstanceState.QUEUED,
                        TaskInstanceState.RUNNING,
                        TaskInstanceState.UP_FOR_RESCHEDULE,
                        TaskInstanceState.UP_FOR_RETRY,
                    )
                )
            )
            .order_by(TI.dag_id, TI.run_id, TI.map_index)
        )
```

mis98zb replied:
Did not see the comment on time...
I'll apply your comments manually into code base.

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -277,9 +284,44 @@ def __get_concurrency_maps(self, states: Iterable[TaskInstanceState], session: S
+            .filter(TI.map_index >= 0)
+            .order_by(TI.dag_id, TI.run_id, TI.map_index)
         )
+        tg_concurrency: dict[tuple[str, str, str], tuple[int, set[int]]] = dict()
```
This type is too complex, add a type alias for it to better explain what the key and value represent.

uranusjr replied:
Also, use `{}` instead of `dict()` (there are multiple instances in the PR)

mis98zb replied:
> Also, use `{}` instead of `dict()` (there are multiple instances in the PR)

applied to the code.

mis98zb replied:
type alias added

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -277,9 +284,44 @@ def __get_concurrency_maps(self, states: Iterable[TaskInstanceState], session: S
+        for (task_id, run_id, dag_id, map_index) in tg_concurrency_query:
+            dag = self.dagbag.get_dag(dag_id, session)
+            task = dag.get_task(task_id) if dag else None
+            if task and task.task_group and task.task_group.max_active_groups_per_dagrun is not None:
```
Better to do something like this to simplify the conditions:

```python
dag = self.dagbag.get_dag(dag_id, session)
if not dag:
    continue
task = dag.get_task(task_id)
if not task:
    continue
task_group = task.task_group
if not task_group:
    continue

mis98zb replied:
updated the code

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -277,9 +284,44 @@ def __get_concurrency_maps(self, states: Iterable[TaskInstanceState], session: S
+                group_id = task.task_group.group_id
+                key = (dag_id, run_id, group_id)
+                if key not in tg_concurrency:
+                    tg_concurrency[key] = (task.task_group.max_active_groups_per_dagrun, set())
```
Maybe `tg_concurrency` can be a defaultdict instead to avoid this

mis98zb replied:
For the first time add a item, it need to set `concurrency limitation` with `task.task_group.max_active_groups_per_dagrun`.
The defaultdict seems cannot set it automatically...

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -513,34 +567,30 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session) -
-                        ).max_active_tis_per_dag
+                    task_dagrun_concurrency_limit: int | None = None
+                    task_group_concurrency_limit: int | None = None
+                    if serialized_dag.has_task(task_id):
+                        serialized_task = serialized_dag.get_task(task_id)
```
Easier to just use `get_task` and catch `TaskNotFound`

mis98zb replied:
I just flowed the legacy code.
Maybe if is faster than try-catch, :)

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -2580,6 +2580,31 @@ def get_num_running_task_instances(self, session: Session, same_dagrun=False) ->
 
+    @provide_session
+    def get_valid_map_index(self, session: Session, limit: int) -> (set[int]):
+        """Return running map index of task group from the DB."""
```
Can logic in this function and TaskGroup be unified?

mis98zb replied:
You mean the code to generate concurrency_map in scheduler_job_runer.py, right?
It should be good if we can unify same logic. But the two places have some differences.

Code here is for task_concurrency_dep.py, which checks for single TI before it be queued.
And the code for concurrency_map is for all TIs after being queued.
So their logic is not very similar.

Commit:
fix style for static checking.

Commit:
update for comments.

Commit:
fix type alias.

Commit:
fix static check again.
I really don't want to check in so many times for this.
But I have trouble in installing breeze on my windows machine, even inside ubutu@wsl2!
I installed mypy but got error when I trying to run it.
What a pittttttttttttttty....

Commit:
oh, one more file left.

Commit:
use BashOperator instead of @task to avoid static check error.

Commit:
remove unused import

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -1401,6 +1401,8 @@ def serialize_task_group(cls, task_group: TaskGroup) -> dict[str, Any] | None:
             }
             encoded["is_mapped"] = True
+            if task_group.max_active_groups_per_dagrun is not None:
+                encoded["max_active_groups_per_dagrun"] = task_group.max_active_groups_per_dagrun
```
Should this not be outside the block instead? Why does this only need to be set for mapped task groups?

mis98zb replied:
If the task group is not mapped, it will always only have one instance per dag run.
In this condition, the concurrency limitation is not needed.

uranusjr replied:
The serialisation data is important for recording purpose though (the result is shown in UI, for example), and silently dropping the value can be confusing.

uranusjr committed:
Merge branch 'main' into task_group_concurrency_control

uranusjr committed:
Move unneeded ti.get_valid_map_index

The function is only used exactly in one place, so the extra abstraction
layer is really not needed, and the logic can be contained entirely in
the TI dep class.

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -277,9 +288,51 @@ def __get_concurrency_maps(self, states: Iterable[TaskInstanceState], session: S
+                        TaskInstanceState.RUNNING,
+                        TaskInstanceState.UP_FOR_RESCHEDULE,
+                        TaskInstanceState.UP_FOR_RETRY,
+                    )
```
Can this use `State.unfinished`, or are there specific reasons some of those states cannot be considered here?

mis98zb replied:
`Sate.unfinished` has four more status: 
- None
- TaskInstanceState.SHUTDOWN,
- TaskInstanceState.RESTARTING,
- TaskInstanceState.DEFERRED,

I'm not sure about their usage.
Let me check.

mis98zb replied:
SHUTDOWN will go FAILED, which is deprecated.
RESTARTING / DEFERRED will be retried.
These three are fine for concurrency limitation.

However, None cannot be used for identify active task group.

So we cannot use `State.unfinished`.

Do we need to add a state list like `State.active`? @uranusjr

uranusjr replied:
None means the task instance has not been touched by the scheduler at all, which is impossible in this context. So I would suggest using `TaskInstance.state.in_(s for s in State.unfinished if s is not None)`

uranusjr committed:
Prevent circular import

uranusjr committed:
Use State.unfinished

uranusjr committed:
Always serialize max_active_groups_per_dagrun

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

Commit:
enhance test case to actually cover multiple queued task instances.

Commit:
enhance test case.

eladkal commented on airflow/decorators/task_group.py:
```diff
@@ -184,6 +184,7 @@ def task_group(
     ui_color: str = "CornflowerBlue",
     ui_fgcolor: str = "#000",
     add_suffix_on_collision: bool = False,
+    max_active_groups_per_dagrun: int | None = None,
```
How would this work with nested task group?

mis98zb replied:
It should be fine since the Operator and task group expansion in an expanded task group is not supported. Because that means we only have one set of map indices.
Let me have a test firstly.

uranusjr replied:
Note that nested expansion may still happen at some point if someone asks for it. It’d be best if this is designed in a way so it leaves enough room to avoid a rewrite in the future.

eladkal commented on tests/jobs/test_scheduler_job.py:
```diff
@@ -1137,6 +1138,142 @@ def test_find_executable_task_instances_max_active_tis_per_dag(self, dag_maker):
+
+            t1 = BashOperator(task_id="dummy", bash_command="sleep 300")
+
+            @task_group(max_active_groups_per_dagrun=concurrency_limit)
```
Please add a test case for nested task group

It's important to test how this behave when `max_active_groups_per_dagrun=1` and the code has task group inside task group:

```
with TaskGroup(group_id='group1') as tg1:
    ...
    with TaskGroup(group_id='group2') as tg2:
    ...
    
```

 We need to make sure it doesn't cause deadlock.

mis98zb replied:
Got it, I'll add the test case.

Since the Operator and task group expansion in an expanded task group is not supported yet (I know this from [here](https://github.com/apache/airflow/discussions/29618#discussioncomment-5027948)), I'll add test cases for 2 scenarios:
1. the expansion on outer group
2. the expansion on inner group

Commit:
more test case to simulate one map index is finished.

mis98zb committed:
Merge branch 'main' into task_group_concurrency_control

Commit:
remove unnecessary change on taskinstance.py

uranusjr commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -86,6 +86,10 @@
 
+# type alias for task concurrency map,
+# (dag_id, run_id, task_group_id) -> concurrency limitation and set of running map indexes
+TaskGroupConcurrencyMap = Dict[Tuple[str, str, str], Tuple[int, Set[int]]]
```
Is the concurrency limit (first int in the value) used anywhere? I tried but could not find one. It seems the code can be simplified a bit if we don’t pass that value.

mis98zb replied:
Yes, you are right.
This field can be removed.

Commit:
avoid test error.

Commit:
remove unnecessary field in TaskGroupConcurrencyMap
